docs/architecture/adr-006-upgrade-client-implementation.md

```diff
@@ -0,0 +1,433 @@
+`latest_height()`, etc.
+
+* Latest client state **MUST NOT** be frozen
+* Latest consensus state **MUST** be within the trusting period of the latest
+  client state
+* Received update message including verification proofs **MUST** successfully be
+  decoded into the domain types
+* Clients **MUST** ensure that the planned last height of the current revision
+  is somehow encoded in the proof verification process. This prevents premature
+  upgrades, as the counterparty may cancel or modify upgrade plans before the
+  last planned height.
+
+Any other requisite beyond the above rules are considered client-specific. Next,
+we go through the upgrade process for Tendermint clients to justify the logic
+and illustrate an example of client-specific upgrade.
+
+### Upgrade Tendermint Clients
+
+This section is based on the `IBC-Go` procedure for client upgrades with a few
+modifications to comply with the `IBC-rs` design patterns.
+
+#### Acceptable Client State Upgrades
+
+Below list enumerates upgrades on Tendermint chains that would break
+counterparty IBC Tendermint clients and indicates whether or not the change is
+supported by `IBC-rs`:
+
+* S: Supported, P: Partially Supported, U: Unsupported
+
+1. (S) Changing the `ChainId`
+2. (P) Changing the `UnbondingPeriod`: chains may increase the unbonding period
+   with no issues. However, decreasing the unbonding period may irreversibly
+   break some counterparty clients. Thus, it is not recommended that chains
+   reduce the unbonding period.
+3. (S) Changing the `Height` (resetting to 0): as long as chains remember to
+   increment the revision number in their chain-id.
+4. (S) Changing the `ProofSpecs`: this should be changed if the proof structure
+   needed to verify IBC proofs is changed across the upgrade. Ex: Switching from
+   an IAVL store, to a SimpleTree Store.
+5. (S) Changing the `UpgradePath`: this might involve changing the key under
+   which upgraded clients and consensus states are stored in the upgrade store,
+   or even migrating the upgrade store itself.
+6. (U) Migrating the IBC store: the store location is negotiated by the
+   connection.
+7. (S) Upgrading to a backwards compatible version of IBC
+8. (U) Upgrading to a non-backwards compatible version of IBC: the version is
+   negotiated on connection handshake.
+9. (U) Changing parameters that are customizable by relayers like `TrustLevel`
+    and `TrustingPeriod`, `max_clock_drift`
+10. (P) Changing the Tendermint LightClient algorithm: Changes to the light
+   client algorithm that do not change the
+   [ClientState](../../crates/ibc/src/clients/ics07_tendermint/client_state.rs#ClientState)
+   or
+   [ConsensusState](../../crates/ibc/src/clients/ics07_tendermint/consensus_state.rs#ConsensusState)
+   struct may be supported, provided that the counterparty is also upgraded to
+   support the new light client algorithm. Changes that require updating the
```

Comment:
It is concerned that the upgrade would change the abstraction of those structs. It shoudn't be so, but is free to change the `ClientState` or `ConsensusState` instances/fields

Reply:
What I'm asking is more: is this a "client upgrade"? The upgrade algorithms are very different:

1. process a `UpgradeClient` message, vs
2. perform a chain hard fork

My understanding was that a client upgrade was only about (1); (2) would be something else

Reply:
You made an accurate point. But, If it would cause a hard fork, it shouldn't be included in the list of possible supported upgrades by IBC-Go. 
Anyway, it is beyond my knowledge at the moment. We should either keep it for the sake of compliance with the IBC-go, but investigate it later?! or discard it and play it by ear if any request comes from our clients. WDYT?

Reply:
Let's keep it for now 👍